pkg/source/reconciler/common/finalizer.go

```diff
@@ -0,0 +1,74 @@
+/*
+Copyright 2021 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package common
+
+import (
+	"context"
+
+	"github.com/Shopify/sarama"
+	"go.uber.org/zap"
+
+	kubeclient "knative.dev/pkg/client/injection/kube/client"
+	"knative.dev/pkg/logging"
+	"knative.dev/pkg/reconciler"
+
+	"knative.dev/eventing-kafka/pkg/apis/sources/v1beta1"
+	"knative.dev/eventing-kafka/pkg/source/client"
+)
+
+const (
+	maxFinalizerAttempts = 10
+)
+
+// BoundedFinalizer limits the number of time a finalizer can run, due to too many errors.
+type BoundedFinalizer struct {
+	FinalizerAttempts map[string]int
+}
+
+func FinalizeKind(ctx context.Context, r *BoundedFinalizer, src *v1beta1.KafkaSource) reconciler.Event {
+	key := src.Namespace + "/" + src.Name
+
+	// Consumer group is an external resource that may not be available anymore so limit the
+	// number of times trying to delete it
```

Comment:
Can you not tell from the error message if the ConsumerGroup is no longer existent and thus not actually pass the error on? That'd spare you all the tracking.

Reply:
I couldn't find this by just reading the doc and didn't get a chance to test it yet. I currently feel the tracking is nice to have since the finalizer is guaranteed to be removed. I'll do more testing and might remove it depending on the findings...